File: core/src/main/java/com/github/fburato/highwheelmodules/core/analysis/LooseAnalyser.java
package com.github.fburato.highwheelmodules.core.analysis;

import com.github.fburato.highwheelmodules.core.algorithms.ModuleGraphTransitiveClosure;
import com.github.fburato.highwheelmodules.model.modules.Definition;
import com.github.fburato.highwheelmodules.model.modules.HWModule;
import com.github.fburato.highwheelmodules.model.modules.ModuleGraph;
import com.github.fburato.highwheelmodules.model.modules.TrackingModuleDependency;
import com.github.fburato.highwheelmodules.model.rules.Dependency;
import com.github.fburato.highwheelmodules.model.rules.NoStrictDependency;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

import static com.github.fburato.highwheelmodules.core.analysis.AnalysisUtils.*;

public class LooseAnalyser implements Analyser {

    @Override
    public AnalyserModel.AnalysisResult analyse(AnalysisState state) {
        return analyseLoose(state);
    }

    private AnalyserModel.AnalysisResult analyseLoose(AnalysisState analysisState) {
        final ModuleGraphTransitiveClosure actualTransitiveClosure = new ModuleGraphTransitiveClosure(
                analysisState.actualGraph, append(analysisState.modules, analysisState.other));

        final List<AnalyserModel.ModuleConnectionViolation> absentDependencyViolations = getAbsentDependencies(
                actualTransitiveClosure, analysisState.dependencies, analysisState.other);
        final List<AnalyserModel.EvidenceBackedViolation> undesiredDependencyViolations = getUndesiredDependecies(
                actualTransitiveClosure, analysisState.noStrictDependencies, analysisState.other,
                analysisState.actualTrackingGraph);

        return new AnalyserModel.AnalysisResult(undesiredDependencyViolations, absentDependencyViolations, getMetrics(
                analysisState.actualGraph, analysisState.modules, analysisState.actualGraph, analysisState.other));
    }

    private List<AnalyserModel.ModuleConnectionViolation> getAbsentDependencies(
            ModuleGraphTransitiveClosure transitiveClosure, Collection<Dependency> dependencies, HWModule other) {
        final List<AnalyserModel.ModuleConnectionViolation> dependencyViolations = new ArrayList<>();
        for (Dependency dependency : dependencies) {
            if (!dependency.source.equals(other) && !dependency.dest.equals(other)
                    && !transitiveClosure.isReachable(dependency.source, dependency.dest)) {
                dependencyViolations
                        .add(new AnalyserModel.ModuleConnectionViolation(dependency.source.name, dependency.dest.name));
            }
        }

        return dependencyViolations;
    }

    private List<AnalyserModel.EvidenceBackedViolation> getUndesiredDependecies(
            ModuleGraphTransitiveClosure transitiveClosure, Collection<NoStrictDependency> noStrictDependencies,
            HWModule other, ModuleGraph<TrackingModuleDependency> trackingGraph) {
        final List<AnalyserModel.EvidenceBackedViolation> undesiredDependencyViolations = new ArrayList<>();
        for (NoStrictDependency noStrictDependency : noStrictDependencies) {
            if (!noStrictDependency.source.equals(other) && !noStrictDependency.dest.equals(other)
                    && transitiveClosure.isReachable(noStrictDependency.source, noStrictDependency.dest)) {
                undesiredDependencyViolations.add(new AnalyserModel.EvidenceBackedViolation(
                        noStrictDependency.source.name, noStrictDependency.dest.name,
                        Arrays.asList(noStrictDependency.source.name, noStrictDependency.dest.name),
                        getNames(transitiveClosure.minimumDistancePath(noStrictDependency.source,
                                noStrictDependency.dest)),
                        getEvidence(trackingGraph, noStrictDependency.source, transitiveClosure
                                .minimumDistancePath(noStrictDependency.source, noStrictDependency.dest))));
            }
        }
        return undesiredDependencyViolations;
    }
}

File: core/src/main/scala/com/github/fburato/highwheelmodules/core/analysis/LooseAnalyser.scala
package com.github.fburato.highwheelmodules.core.analysis

import com.github.fburato.highwheelmodules.core.algorithms.ModuleGraphTransitiveClosure
import com.github.fburato.highwheelmodules.model.modules.{HWModuleS, ModuleGraphS, TrackingModuleDependencyS}
import com.github.fburato.highwheelmodules.model.rules.{DependencyS, NoStrictDependencyS}

private[analysis] object LooseAnalyser extends Analyser {
  override def analyse(state: AnalysisState): AnalysisResult = {
    val actualTransitiveClosure = ModuleGraphTransitiveClosure(state.actualGraph, (state.modules ++ Seq(state.other)))
    val absentDependencies = getAbsentDependencies(actualTransitiveClosure, state.dependencies, state.other)
    val undesiredDependencies = getUndesiredDependencies(actualTransitiveClosure, state.noStrictDependencies, state.other, state.actualTrackingGraph)

    AnalysisResult(undesiredDependencies,
      absentDependencies, getMetrics(state.actualGraph, state.modules, state.actualGraph, state.other))
  }

  private def getAbsentDependencies(transitiveClosure: ModuleGraphTransitiveClosure,
                                    dependencies: Seq[DependencyS],
                                    other: HWModuleS): Seq[ModuleConnectionViolation] = {
    dependencies
      .filter(d => d.source != other && d.dest != other && !transitiveClosure.isReachable(d.source, d.dest))
      .map(d => ModuleConnectionViolation(d.source.name, d.dest.name))
  }

  private def getUndesiredDependencies(transitiveClosure: ModuleGraphTransitiveClosure,
                                       noStrictDependencies: Seq[NoStrictDependencyS],
                                       other: HWModuleS,
                                       trackingGraph: ModuleGraphS[TrackingModuleDependencyS]): Seq[EvidenceBackedViolation] = {
    noStrictDependencies
      .filter(d => d.source != other && d.dest != other && transitiveClosure.isReachable(d.source, d.dest))
      .map(d => EvidenceBackedViolation(d.source.name, d.dest.name,
        List(d.source.name, d.dest.name),
        getNames(transitiveClosure.minimumDistancePath(d.source, d.dest)),
        getEvidence(trackingGraph, d.source, transitiveClosure.minimumDistancePath(d.source, d.dest))))
  }
}
